### ISSUE TITLE
MemberNotNullWhen does not work when overridden in derived interface

### ISSUE THREAD

**Issue by Ilchert** [dotnet/roslyn#76743]
**Version Used**: 
SDK 9.0.101
**Steps to Reproduce**:

1. Mark overridden property as MemberNotNullWhen from base interface
2.  Roslyn ignores this attribute

[sharplab](https://sharplab.io/#v2:D4AQDABCCMB0AiBLAhgcwHYHsDOAXRAxtrAMKYAmApgILrIA2AntotgNwCwAUCAEzTduAN2QAnCOIC8EdJQDuEAEqVsAB0zpslADwwwAPgAUAInqZ6xgDQRcogK6UAlJx7QAnIYDilXIdGPnQS4oaEhvXwBJZTUNLV1Q/QhsRwgAb24ITKzMxAAzQ2II7ABlOwICFWSM7JrMkAB2JNINXEp0XBdausaIYypc5Dt6XGNOrIBfIO4+JRV1TR0AFSNFiAIWttxrACNMcySyiuwqrgAuCCi52KX9bnTgrJAAZghVsnbNiElE9Y/2l2qdReu32RVK5UqX0S2EOlRcky4QUQH1EAwqF2i8y0d0BmRB9AuJVhxzSqB8bARCO4yNaqOQ6MuMQW2mWp0ZWMoOIemQA2gBZSgAW22lFEADlMLgxUN6AB1AAWbUMtgcljogsomHy71a7QCAF1cRBkNs8KJ6bgIPiMVcFrAwcTsKTyQiaosAPwQHWfVJkjqUoA==)

```C#
string Get(IResponse<string> s) {
        if(s.IsSuccess)
            return s.Content; // warning CS8603: Possible null reference return.
        return  "default";
    }

interface IResponse
{
     bool IsSuccess {get;}
}

interface IResponse<T>:IResponse
{
    [MemberNotNullWhen(true,nameof(Content))]
     abstract bool IResponse.IsSuccess {get;}
     T? Content {get;}
}

```

**Diagnostic Id**:
warning CS8603: Possible null reference return.

**Expected Behavior**:
Should not trigger warning.

**Actual Behavior**:
As far I understand MemberNotNullWhen can reference only properties/methods from same class. Like in the example. But in this case, it ignores the attribute on IsSuccess  override.

**Comment by RikkiGibson** [comment]
I think when the `IsSuccess` declaration is an explicit interface implementation, that member is not "seen" at the use site. Only the signature of the implemented member is seen. So if you want attributes and such to have an effect at the use site you may want to do something like this: [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgPgAQAwAIEEYB0ARKEDmMA9gM4AuUAxiZgMJEAmApgIIwQA2AniVCQNwBYAFAiMqAOJMyACgCSAJSYkADkRgkmAHnEA+FCQCUKAN4iUFyxagAzGTTkkAygFdKlZUfNWfFhAHYDOnUyJhgyfhQAeiiUIgBrb18AiwAiZhsIFw4yVKFhKwBfERFYUIAnTI8URWU1DSYRMwKrYCIiDhrnNw8SElN8aX5i4RHS8KZKiGra1XVNLQAVXRBZ+s0mpJQ2jpqlOYbMR1d3TxQAXn1jns98rYBtAFkmMGBJgDkiMnfsjgB1AAWYRkZHKLiYABp2GAmEQ7PQJuFDIYALpbFAQYDkcrTMgoGBMADu23anWupz6AyGIx8iwA/CgEaFwlSIiNCkA==)

```cs
using System.Diagnostics.CodeAnalysis;

string Get(IResponse<string> s) {
        if(s.IsSuccess)
            return s.Content; // ok
        return  "default";
    }

interface IResponse
{
     bool IsSuccess {get;}
}

interface IResponse<T>:IResponse
{
    bool IResponse.IsSuccess => IsSuccess;

    [MemberNotNullWhen(true,nameof(Content))]
     abstract new bool IsSuccess {get;}
     T? Content {get;}
}
```

**Comment by RikkiGibson** [comment]
Closing out as answered. Feel free to follow up with any more questions or concerns. Thanks!

**Comment by Ilchert** [comment]
Hello @RikkiGibson. Thanks for your answer, but your solution creates new members, which leads to problems with inheritance.

```C#

var m = new Mock<IResponse<string>>();
m.Setup(p => p.IsSuccess).Returns(true);
m.Setup(p => p.Content).Returns("moq");

Console.WriteLine(Get(m.Object));
Console.WriteLine(GetSuccess(m.Object));


string Get(IResponse<string> s)
{
    if (s.IsSuccess)
        return s.Content; // ok
    return "default";
}

bool GetSuccess(IResponse s)
{
    return s.IsSuccess;
}

interface IResponse
{
    bool IsSuccess { get; }
}

interface IResponse<T> : IResponse
{
    bool IResponse.IsSuccess => IsSuccess;

    [MemberNotNullWhen(true, nameof(Content))]
    abstract new bool IsSuccess { get; }
    T? Content { get; }
}
```
This code will return
```console
moq
False
```

And it is unexpected for developers. Also using default interface implementation can affect structs. Original example works fine with Moq.

Thanks, Ilya.
